File: leet-code/209.长度最小的子数组.js
/*
 * @lc app=leetcode.cn id=209 lang=javascript
 *
 * [209] 长度最小的子数组
 */

// @lc code=start
/**
 * @param {number} target
 * @param {number[]} nums
 * @return {number}
 */
var minSubArrayLen = function(target, nums) {
  let left = 0, right = 0, sum = 0, res = Infinity;
  while (right < nums.length) {
    sum += nums[right];
    while (sum >= target) {
      res = Math.min(res, right - left + 1);
      sum -= nums[left];
      left++;
    }
    right++;
  }
  return res === Infinity ? 0 : res;
};
// @lc code=end

File: leet-code-rs/src/solutions/209_长度最小的子数组.rs
/*
 * @lc app=leetcode.cn id=209 lang=rust
 *
 * [209] 长度最小的子数组
 */

// @lc code=start
impl Solution {
    pub fn min_sub_array_len(target: i32, nums: Vec<i32>) -> i32 {
        let mut slow = 0;
        let mut sum = 0;
        let mut res = usize::MAX;
        for fast in 0..nums.len() {
            sum += nums[fast];
            // 缩短数组长度
            while sum >= target {
                if sum - nums[slow] >= target {
                    sum -= nums[slow];
                    slow += 1;
                } else {
                    res = res.min(fast - slow + 1);
                    break;
                }
            }
        }

        return if res == usize::MAX { 0 } else { res } as i32;
    }
}
// @lc code=end

#[allow(unused)]
struct Solution;

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_min_sub_array_len() {
        assert_eq!(Solution::min_sub_array_len(7, vec![2, 3, 1, 2, 4, 3]), 2);
        assert_eq!(Solution::min_sub_array_len(3, vec![1, 4, 4]), 1);
        assert_eq!(
            Solution::min_sub_array_len(11, vec![1, 1, 1, 1, 1, 1, 1, 1]),
            0
        );
        assert!(true)
    }
}
